File: dynamic/count_construct.py
## return the number of ways the word bank can create the target string
## without memoization: 0(n^m)
## with memo: 0(n * m^2)
def count_construct(target, word_bank, memo = {}):
  if target in memo: 
    return memo[target]
  if target == "":
    return 1
  total_count = 0
  for word in word_bank: 
    # startwith built method return True if valid prefix
    if target.startswith(word):
      num_ways = count_construct(target[len(word):], word_bank)
      total_count += num_ways
  memo[target] = total_count
  return total_count

  




print(count_construct("abcdef", ["ab", "abc", "cd", "def", "abcd"]))

File: dynamic/countConstruct.ts
const countConstruct = (target: string, wordBank: string[]): number => {
  const dp = new Array(target.length + 1).fill(0);
  dp[0] = 1;
  for (let i = 0; i <= target.length; i++) {
    if (dp[i] > 0) {
      for (let word of wordBank) {
        if (word === target.slice(i, i + word.length)) {
          dp[i + word.length] += dp[i];
        }
      }
    }
  }

  return dp[target.length];
};

//  number to right represents num of ways to contruct substring
// so   "purp" has 2 ways to be contructed.
//  [1, 1, 0, 1, 2, 1, 2]
//   p  u  r  p  l  e  empty

console.log(countConstruct("purple", ["purp", "p", "ur", "le", "purpl"]));
